tools/subtype-gen/rules.yaml

```diff
@@ -0,0 +1,433 @@
+# $schema: ./rules.schema.json
+#
+# Cadence - The resource-oriented smart contract programming language
+#
+# Copyright Flow Foundation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+rules:
+  - super: AnyType
+    predicate:
+      always
+
+  - super: AnyStructType
+    predicate:
+      and:
+        - not:
+            isResource: sub
+        - not:
+            equals:
+              source: sub
+              target: AnyType
+
+  - super: AnyResourceType
+    predicate:
+      isResource: sub
+
+  - super: AnyResourceAttachmentType
+    predicate:
+      and:
+        - isAttachment: sub
+        - isResource: sub
+
+  - super: AnyStructAttachmentType
+    predicate:
+      and:
+        - isAttachment: sub
+        - not:
+            isResource: sub
+
+  - super: HashableStructType
+    predicate:
+      isHashableStruct: sub
+
+  - super: PathType
+    predicate:
+      subtype:
+        sub: sub
+        super:
+          # Maybe use 'or' predicate here instead?
+          # The difference is 'or' is a predicate containing zero or more inner predicates.
+          # Whereas, 'oneOf' contains zero or more values, not predicates.
+          oneOf: [StoragePathType, CapabilityPathType]
```

Comment:
I only noticed this comment in the code once I had already left the comment below. Agreed that using `or` might be simpler.